File: teambadass/utils/storage_manager.py
#!/usr/bin/env python3
"""
TeamBadass Storage Manager
Unified interface for repository, external and cloud storage
"""

import os
import json
import logging
import shutil
from datetime import datetime
from pathlib import Path

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(message)s')
logger = logging.getLogger("StorageManager")

class StorageManager:
    """Manages external storage locations and operations"""
    
    def __init__(self, config_path=None):
        """Initialize with optional config path"""
        self.repo_path = "/home/louthenw/Documents/teambadass_github/teambadass"
        
        # Default external storage location (outside repo)
        self.external_path = "/home/louthenw/teambadass_external"
        
        # Load config if available
        self.config = self._load_config(config_path)
        
        # Apply configuration
        self._apply_config()
    
    def _load_config(self, config_path=None):
        """Load configuration from file"""
        # Default config path
        if not config_path:
            config_path = os.path.join(self.repo_path, "memory", "config", "storage.json")
        
        # Try to load config
        try:
            if os.path.exists(config_path):
                with open(config_path, 'r') as f:
                    return json.load(f)
        except Exception as e:
            logger.warning(f"Could not load config: {e}")
        
        # Return default config
        return {
            "external_path": self.external_path,
            "backup_path": os.path.join(self.external_path, "backups"),
            "temp_path": os.path.join(self.external_path, "temp"),
            "archive_path": os.path.join(self.external_path, "archive"),
            "cloud_enabled": False,
            "cloud_provider": None,
            "cloud_config": {}
        }
    
    def _apply_config(self):
        """Apply configuration settings"""
        # Set paths from config
        self.external_path = self.config.get("external_path", self.external_path)
        self.backup_path = self.config.get("backup_path", os.path.join(self.external_path, "backups"))
        self.temp_path = self.config.get("temp_path", os.path.join(self.external_path, "temp"))
        self.archive_path = self.config.get("archive_path", os.path.join(self.external_path, "archive"))
        
        # Ensure directories exist
        for path in [self.external_path, self.backup_path, self.temp_path, self.archive_path]:
            os.makedirs(path, exist_ok=True)
    
    def save_config(self, config_path=None):
        """Save current configuration"""
        if not config_path:
            config_path = os.path.join(self.repo_path, "memory", "config", "storage.json")
        
        # Ensure directory exists
        os.makedirs(os.path.dirname(config_path), exist_ok=True)
        
        # Save config
        with open(config_path, 'w') as f:
            json.dump(self.config, f, indent=2)
        
        logger.info(f"Configuration saved to: {config_path}")
        return True
    
    def backup_file(self, source_path, category="general"):
        """Backup a file to external storage"""
        # Normalize source path
        if not os.path.isabs(source_path):
            source_path = os.path.join(self.repo_path, source_path)
        
        # Ensure source exists
        if not os.path.exists(source_path):
            logger.error(f"Source file not found: {source_path}")
            return False
        
        # Create category directory
        category_dir = os.path.join(self.backup_path, category)
        os.makedirs(category_dir, exist_ok=True)
        
        # Create destination path
        filename = os.path.basename(source_path)
        dest_path = os.path.join(category_dir, filename)
        
        # Copy file
        try:
            shutil.copy2(source_path, dest_path)
            logger.info(f"Backed up: {source_path} → {dest_path}")
            return dest_path
        except Exception as e:
            logger.error(f"Backup failed: {e}")
            return False
    
    def create_external_directory(self, directory, exist_ok=True):
        """Create directory in external storage"""
        path = os.path.join(self.external_path, directory)
        os.makedirs(path, exist_ok=exist_ok)
        logger.info(f"Created external directory: {path}")
        return path
    
    def move_to_external(self, source_path, dest_dir=None, remove_original=True):
        """Move file or directory to external storage"""
        # Normalize source path
        if not os.path.isabs(source_path):
            source_path = os.path.join(self.repo_path, source_path)
        
        # Ensure source exists
        if not os.path.exists(source_path):
            logger.error(f"Source not found: {source_path}")
            return False
        
        # Determine destination directory
        if not dest_dir:
            dest_dir = ""
        
        dest_path = os.path.join(self.external_path, dest_dir)
        os.makedirs(dest_path, exist_ok=True)
        
        # Get destination filename
        filename = os.path.basename(source_path)
        full_dest_path = os.path.join(dest_path, filename)
        
        # Perform move/copy
        try:
            if remove_original:
                shutil.move(source_path, full_dest_path)
                logger.info(f"Moved: {source_path} → {full_dest_path}")
            else:
                if os.path.isfile(source_path):
                    shutil.copy2(source_path, full_dest_path)
                else:
                    shutil.copytree(source_path, full_dest_path)
                logger.info(f"Copied: {source_path} → {full_dest_path}")
            
            return full_dest_path
        except Exception as e:
            logger.error(f"Operation failed: {e}")
            return False
    
    def archive_directory(self, source_path, archive_name=None):
        """Archive a directory to the archive location"""
        # Normalize source path
        if not os.path.isabs(source_path):
            source_path = os.path.join(self.repo_path, source_path)
        
        # Ensure source exists
        if not os.path.exists(source_path):
            logger.error(f"Source not found: {source_path}")
            return False
        
        # Generate archive name if not provided
        if not archive_name:
            dirname = os.path.basename(source_path)
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            archive_name = f"{dirname}_{timestamp}"
        
        # Create archive path
        archive_path = os.path.join(self.archive_path, archive_name)
        
        # Move directory to archive
        try:
            shutil.move(source_path, archive_path)
            logger.info(f"Archived: {source_path} → {archive_path}")
            return archive_path
        except Exception as e:
            logger.error(f"Archive failed: {e}")
            return False
    
    def get_external_path(self, relative_path=None):
        """Get absolute path in external storage"""
        if not relative_path:
            return self.external_path
        
        return os.path.join(self.external_path, relative_path)
    
    def list_external_directory(self, relative_path=None):
        """List contents of external directory"""
        path = self.get_external_path(relative_path)
        
        if not os.path.exists(path):
            logger.error(f"Path not found: {path}")
            return []
        
        try:
            return os.listdir(path)
        except Exception as e:
            logger.error(f"Error listing directory: {e}")
            return []

# Run tests if executed directly
if __name__ == "__main__":
    # Create storage manager
    manager = StorageManager()
    
    # Display paths
    print(f"TeamBadass Storage Manager")
    print(f"===========================")
    print(f"Repository Path: {manager.repo_path}")
    print(f"External Path: {manager.external_path}")
    print(f"Backup Path: {manager.backup_path}")
    print(f"Temp Path: {manager.temp_path}")
    print(f"Archive Path: {manager.archive_path}")
    
    # Display cloud status
    cloud_enabled = manager.config.get("cloud_enabled", False)
    cloud_provider = manager.config.get("cloud_provider")
    print(f"\nCloud Storage: {'Enabled' if cloud_enabled else 'Disabled'}")
    if cloud_enabled and cloud_provider:
        print(f"Cloud Provider: {cloud_provider}")
    
    # List external directories
    print(f"\nExternal Directories:")
    for directory in ["backups", "temp", "archive"]:
        path = manager.get_external_path(directory)
        if os.path.exists(path):
            items = manager.list_external_directory(directory)
            print(f"- {directory}: {len(items)} items")

File: teambadass/utils/storage/storage-manager.js
/**
 * TeamBadass Storage Manager
 * Provides management for external storage locations and operations
 * 
 * FILE_OVERVIEW: Flexible storage management with local and cloud capabilities
 * VERSION: 1.0.0
 * LAST_UPDATED: 2025-05-03
 * DEPENDENCIES: fs-helper.js
 * 
 * TABLE_OF_CONTENTS:
 * 1. StorageManager Class - Main storage management implementation
 * 2. Configuration Management - Loading and saving storage configuration
 * 3. Directory Operations - Creating and validating storage directories
 * 4. File Operations - Moving and copying files to external storage
 * 5. Backup Operations - Creating and managing backups
 * 
 * SKIP_DETAILED_ANALYSIS: True - This header provides sufficient context
 */

const path = require('path');
const fsHelper = require('../fs-helper');

/**
 * Storage Manager for TeamBadass
 * Manages external storage locations and operations
 */
class StorageManager {
  /**
   * Create a new StorageManager
   * @param {Object} options - Configuration options
   * @param {string} options.repoPath - Repository base path
   * @param {string} options.externalPath - External storage base path
   * @param {boolean} options.verbose - Enable verbose logging
   */
  constructor(options = {}) {
    // Default options
    this.options = {
      repoPath: '/home/louthenw/Documents/teambadass_github/teambadass',
      externalPath: '/home/louthenw/teambadass_external',
      verbose: true,
      ...options
    };

    // Storage paths
    this.repoPath = this.options.repoPath;
    this.externalPath = this.options.externalPath;
    this.configPath = path.join(this.repoPath, 'memory', 'config', 'storage.json');

    // Additional paths will be set after loading config
    this.backupPath = null;
    this.tempPath = null;
    this.archivePath = null;

    // Storage state
    this.initialized = false;
    this.config = null;
    this.cloudEnabled = false;
  }

  /**
   * Initialize the storage manager
   * @returns {Promise<Object>} Initialization result
   */
  async initialize() {
    try {
      // Load configuration
      await this.loadConfig();

      // Ensure directories exist
      await this.ensureDirectories();

      // Create symbolic link if possible
      await this.createSymLink();

      this.initialized = true;
      this.log('Storage manager initialized successfully');

      return {
        status: 'success',
        paths: {
          external: this.externalPath,
          backup: this.backupPath,
          temp: this.tempPath,
          archive: this.archivePath
        },
        cloudEnabled: this.cloudEnabled
      };
    } catch (error) {
      this.log(`Initialization failed: ${error.message}`, 'error');
      return {
        status: 'error',
        message: error.message
      };
    }
  }

  /**
   * Load configuration from file or create default
   * @returns {Promise<Object>} Configuration object
   */
  async loadConfig() {
    try {
      // Check if config exists
      const configExists = await fsHelper.pathExists(this.configPath);

      if (configExists) {
        // Load existing config
        const configContent = await fs.read_file({ path: this.configPath });
        this.config = JSON.parse(configContent);
        this.log('Configuration loaded from file');
      } else {
        // Create default config
        this.config = {
          external_path: this.externalPath,
          backup_path: path.join(this.externalPath, 'backups'),
          temp_path: path.join(this.externalPath, 'temp'),
          archive_path: path.join(this.externalPath, 'archive'),
          cloud_enabled: false,
          cloud_provider: null,
          cloud_config: {}
        };
        this.log('Created default configuration');
      }

      // Apply configuration
      this.externalPath = this.config.external_path;
      this.backupPath = this.config.backup_path;
      this.tempPath = this.config.temp_path;
      this.archivePath = this.config.archive_path;
      this.cloudEnabled = this.config.cloud_enabled;

      return this.config;
    } catch (error) {
      this.log(`Error loading configuration: ${error.message}`, 'error');
      throw error;
    }
  }

  /**
   * Save configuration to file
   * @returns {Promise<boolean>} Success status
   */
  async saveConfig() {
    try {
      // Ensure directories exist
      await fsHelper.ensureDirectory(path.dirname(this.configPath));

      // Save config
      await fsHelper.writeFile(this.configPath, JSON.stringify(this.config, null, 2));
      this.log('Configuration saved to file');

      return true;
    } catch (error) {
      this.log(`Error saving configuration: ${error.message}`, 'error');
      return false;
    }
  }

  /**
   * Ensure all required directories exist
   * @returns {Promise<boolean>} Success status
   */
  async ensureDirectories() {
    try {
      // Create directories
      await fsHelper.ensureDirectory(this.externalPath);
      await fsHelper.ensureDirectory(this.backupPath);
      await fsHelper.ensureDirectory(this.tempPath);
      await fsHelper.ensureDirectory(this.archivePath);

      this.log('Storage directories created');
      return true;
    } catch (error) {
      this.log(`Error creating directories: ${error.message}`, 'error');
      throw error;
    }
  }

  /**
   * Create symbolic link to external storage
   * @returns {Promise<boolean>} Success status
   */
  async createSymLink() {
    try {
      // Skip if not running in an environment that supports symlinks
      if (typeof process === 'undefined') {
        return false;
      }

      const linkPath = path.join(this.repoPath, '.external');

      // Create symlink
      // Note: This would need NodeJS environment to work
      // In a real implementation, this would use fs.symlink
      this.log(`Would create symlink: ${linkPath} -> ${this.externalPath}`);

      return true;
    } catch (error) {
      this.log(`Error creating symlink: ${error.message}`, 'warning');
      return false;
    }
  }

  /**
   * Create a backup of a file or directory
   * @param {string} sourcePath - Path to file or directory to backup
   * @param {string} category - Category for organization (optional)
   * @returns {Promise<Object>} Backup result
   */
  async backup(sourcePath, category = 'general') {
    if (!this.initialized) {
      await this.initialize();
    }

    try {
      // Normalize path
      if (!path.isAbsolute(sourcePath)) {
        sourcePath = path.join(this.repoPath, sourcePath);
      }

      // Ensure source exists
      const sourceExists = await fsHelper.pathExists(sourcePath);
      if (!sourceExists) {
        throw new Error(`Source not found: ${sourcePath}`);
      }

      // Create category directory
      const categoryPath = path.join(this.backupPath, category);
      await fsHelper.ensureDirectory(categoryPath);

      // Get timestamp
      const timestamp = new Date().toISOString().replace(/:/g, '-').replace(/\..+/, '');
      
      // Get basename
      const baseName = path.basename(sourcePath);
      
      // Create destination path with timestamp
      const destName = `${baseName}_${timestamp}`;
      const destPath = path.join(categoryPath, destName);

      // Copy file or directory
      // In a real implementation, this would use the appropriate fs methods
      this.log(`Backing up ${sourcePath} to ${destPath}`);

      // For demonstration, we'll just write a placeholder file
      await fsHelper.writeFile(
        destPath + '.info',
        JSON.stringify({
          originalPath: sourcePath,
          backupTime: timestamp,
          category: category
        }, null, 2)
      );

      return {
        status: 'success',
        originalPath: sourcePath,
        backupPath: destPath,
        timestamp: timestamp
      };
    } catch (error) {
      this.log(`Backup failed: ${error.message}`, 'error');
      return {
        status: 'error',
        message: error.message
      };
    }
  }

  /**
   * Move a file or directory to external storage
   * @param {string} sourcePath - Path to file or directory
   * @param {string} destDir - Destination directory (relative to external path)
   * @param {boolean} keepOriginal - Keep original if true, move if false
   * @returns {Promise<Object>} Operation result
   */
  async moveToExternal(sourcePath, destDir = '', keepOriginal = false) {
    if (!this.initialized) {
      await this.initialize();
    }

    try {
      // Normalize path
      if (!path.isAbsolute(sourcePath)) {
        sourcePath = path.join(this.repoPath, sourcePath);
      }

      // Ensure source exists
      const sourceExists = await fsHelper.pathExists(sourcePath);
      if (!sourceExists) {
        throw new Error(`Source not found: ${sourcePath}`);
      }

      // Create destination directory
      const fullDestDir = path.join(this.externalPath, destDir);
      await fsHelper.ensureDirectory(fullDestDir);

      // Get basename
      const baseName = path.basename(sourcePath);
      
      // Full destination path
      const destPath = path.join(fullDestDir, baseName);

      // Copy or move
      const operation = keepOriginal ? 'copy' : 'move';
      this.log(`${operation} ${sourcePath} to ${destPath}`);

      // For demonstration, we'll just write a placeholder file
      await fsHelper.writeFile(
        destPath + '.info',
        JSON.stringify({
          originalPath: sourcePath,
          operation: operation,
          timestamp: new Date().toISOString()
        }, null, 2)
      );

      return {
        status: 'success',
        operation: operation,
        originalPath: sourcePath,
        destinationPath: destPath
      };
    } catch (error) {
      this.log(`${keepOriginal ? 'Copy' : 'Move'} operation failed: ${error.message}`, 'error');
      return {
        status: 'error',
        message: error.message
      };
    }
  }

  /**
   * Archive a directory to external storage
   * @param {string} sourcePath - Path to directory
   * @param {string} name - Archive name (optional)
   * @returns {Promise<Object>} Archive result
   */
  async archiveDirectory(sourcePath, name = '') {
    if (!this.initialized) {
      await this.initialize();
    }

    try {
      // Normalize path
      if (!path.isAbsolute(sourcePath)) {
        sourcePath = path.join(this.repoPath, sourcePath);
      }

      // Ensure source exists
      const sourceExists = await fsHelper.pathExists(sourcePath);
      if (!sourceExists) {
        throw new Error(`Source not found: ${sourcePath}`);
      }

      // Generate archive name if not provided
      const baseName = name || path.basename(sourcePath);
      const timestamp = new Date().toISOString().replace(/:/g, '-').replace(/\..+/, '');
      const archiveName = `${baseName}_${timestamp}`;
      
      // Create archive directory
      const archivePath = path.join(this.archivePath, archiveName);
      await fsHelper.ensureDirectory(archivePath);
      
      this.log(`Archiving ${sourcePath} to ${archivePath}`);

      // For demonstration, we'll just write a placeholder file
      await fsHelper.writeFile(
        archivePath + '.info',
        JSON.stringify({
          originalPath: sourcePath,
          archiveTime: timestamp,
          name: archiveName
        }, null, 2)
      );

      return {
        status: 'success',
        originalPath: sourcePath,
        archivePath: archivePath,
        name: archiveName
      };
    } catch (error) {
      this.log(`Archive operation failed: ${error.message}`, 'error');
      return {
        status: 'error',
        message: error.message
      };
    }
  }

  /**
   * Create a temporary file or directory
   * @param {string} name - Name for temp file/directory
   * @param {string} content - Content for file (optional)
   * @returns {Promise<Object>} Operation result
   */
  async createTemp(name, content = '') {
    if (!this.initialized) {
      await this.initialize();
    }

    try {
      // Generate unique name if not provided
      const tempName = name || `temp_${Date.now()}`;
      
      // Full temp path
      const tempPath = path.join(this.tempPath, tempName);
      
      // Write content if provided
      if (content) {
        await fsHelper.writeFile(tempPath, content);
        this.log(`Created temp file: ${tempPath}`);
      } else {
        await fsHelper.ensureDirectory(tempPath);
        this.log(`Created temp directory: ${tempPath}`);
      }

      return {
        status: 'success',
        path: tempPath,
        name: tempName,
        type: content ? 'file' : 'directory'
      };
    } catch (error) {
      this.log(`Temp operation failed: ${error.message}`, 'error');
      return {
        status: 'error',
        message: error.message
      };
    }
  }

  /**
   * Configure cloud storage
   * @param {string} provider - Cloud provider ('dropbox', 'google', etc.)
   * @param {Object} config - Provider-specific configuration
   * @returns {Promise<Object>} Configuration result
   */
  async configureCloud(provider, config = {}) {
    if (!this.initialized) {
      await this.initialize();
    }

    try {
      // Validate provider
      const validProviders = ['dropbox', 'google', 'onedrive', 'custom'];
      if (!validProviders.includes(provider.toLowerCase())) {
        throw new Error(`Unsupported cloud provider: ${provider}`);
      }

      // Update configuration
      this.config.cloud_enabled = true;
      this.config.cloud_provider = provider.toLowerCase();
      this.config.cloud_config = config;
      
      // Save configuration
      await this.saveConfig();
      
      this.cloudEnabled = true;
      this.log(`Cloud storage configured: ${provider}`);

      return {
        status: 'success',
        provider: provider,
        enabled: true
      };
    } catch (error) {
      this.log(`Cloud configuration failed: ${error.message}`, 'error');
      return {
        status: 'error',
        message: error.message
      };
    }
  }

  /**
   * Log a message
   * @param {string} message - Message to log
   * @param {string} level - Log level ('info', 'warning', 'error')
   */
  log(message, level = 'info') {
    if (!this.options.verbose && level === 'info') {
      return;
    }

    const prefix = {
      info: '📢 INFO:',
      warning: '⚠️ WARNING:',
      error: '❌ ERROR:',
    }[level] || '   ';

    console.log(`${prefix} ${message}`);
  }
}

module.exports = StorageManager;
